sign/rpmsignverity.c

```diff
@@ -0,0 +1,234 @@
+/**
+ * Copyright (C) 2020 Facebook
+ *
+ * Author: Jes Sorensen <jsorensen@fb.com>
+ */
+
+#include "system.h"
+
+#include <rpm/rpmlib.h>		/* RPMSIGTAG & related */
+#include <rpm/rpmlog.h>		/* rpmlog */
+#include <rpm/rpmfi.h>
+#include <rpm/rpmpgp.h>		/* rpmDigestLength */
+#include "lib/header.h"		/* HEADERGET_MINMEM */
+#include "lib/header_internal.h"
+#include "lib/rpmtypes.h"	/* rpmRC */
+#include <libfsverity.h>
+#include <rpmio/rpmio_internal.h>
+#include "lib/rpmvs.h"
+
+#include "sign/rpmsignverity.h"
+
+#define MAX_SIGNATURE_LENGTH 1024
+
+static int rpmVerityRead(void *opaque, void *buf, size_t size)
+{
+	int retval;
+	rpmfi fi = (rpmfi)opaque;
+
+	retval = rpmfiArchiveRead(fi, buf, size);
+
+	if (retval > 0)
+		retval = 0;
+	return retval;
+}
+
+static char *rpmVeritySignFile(rpmfi fi, size_t *sig_size, char *key,
+			       char *keypass, char *cert, uint16_t algo,
+			       uint32_t block_size)
+{
+    struct libfsverity_merkle_tree_params params;
+    struct libfsverity_signature_params sig_params;
+    struct libfsverity_digest *digest = NULL;
+    rpm_loff_t file_size;
+    char *digest_hex, *sig_hex = NULL;
+    uint8_t *sig = NULL;
+    int status;
+
+    if (S_ISLNK(rpmfiFMode(fi)))
+	file_size = 0;
+    else
+	file_size = rpmfiFSize(fi);
+
+    memset(&params, 0, sizeof(struct libfsverity_merkle_tree_params));
+    params.version = 1;
+    params.hash_algorithm = algo;
+    params.block_size = block_size;
+    params.salt_size = 0 /* salt_size */;
+    params.salt = NULL /* salt */;
+    status = libfsverity_compute_digest(fi, file_size, rpmVerityRead,
+					&params, &digest);
+    if (status) {
+	rpmlog(RPMLOG_DEBUG, _("failed to compute digest\n"));
+	goto out;
+    }
+
+    digest_hex = pgpHexStr(digest->digest, digest->digest_size);
+    rpmlog(RPMLOG_DEBUG, _("file(size %li): %s: digest(%i): %s, idx %i\n"),
+	   file_size, rpmfiFN(fi), digest->digest_size, digest_hex,
+	   rpmfiFX(fi));
+
+    free(digest_hex);
+
+    memset(&sig_params, 0, sizeof(struct libfsverity_signature_params));
+    sig_params.keyfile = key;
+    sig_params.certfile = cert;
+    if (libfsverity_sign_digest(digest, &sig_params, &sig, sig_size)) {
+	rpmlog(RPMLOG_DEBUG, _("failed to sign digest\n"));
+	goto out;
+    }
+
+    sig_hex = pgpHexStr(sig, *sig_size);
```

Comment:
Here too, IMA file signatures set a bad example. Use base64 encoded strings instead of hex, it's much more space efficient (and IMA should be changed to that as well)

Reply:
Good point, I used IMA file signatures as a reference since it's the closest in behavior to what I need. Switching to base64 makes a ton of sense though, so I'll update the code to use that instead, and push an update when I have that working.